core/shared/src/main/scala/com/lynbrookrobotics/potassium/streams/Stream.scala

```diff
@@ -20,38 +16,80 @@ abstract class Stream[+T] { self =>
 
   private[this] var listeners = Vector.empty[T => Unit]
 
+  private var hasShutdown = false
+  def subscribeToParents(): Unit
+  def unsubscribeFromParents(): Unit
+  def checkRelaunch(): Unit = {}
+
   protected def publishValue(value: T @uncheckedVariance): Unit = {
     listeners.foreach(_.apply(value))
     // TODO: more stuff maybe
   }
 
+  /**
+    * Adds a listener for elements of this stream. Callbacks will be executed
+    * whenever a new value is published in order of when the callbacks were added.
+    * Callbacks added first will be called first and callbacks added last will
+    * be called last.
+    *
+    * @param thunk the listener to be called on each published value
+    * @return a function to call to remove the listener
+    */
+  def foreach(thunk: T => Unit): Cancel = {
+    self.synchronized {
+      listeners = listeners :+ thunk
+      if (listeners.size == 1) {
+        // first listener, boot up!
+        if (hasShutdown) {
+          checkRelaunch()
+        }
+
+        subscribeToParents()
+      }
+    }
+
+    new Cancel {
+      override def cancel(): Unit = self.synchronized {
+        val beforeSize = listeners.size
+        listeners = listeners.filterNot(_ eq thunk)
+
+        if (beforeSize > 0 && listeners.isEmpty) {
+          // we are shutting down!
+          unsubscribeFromParents()
+          hasShutdown = true
+        }
+      }
+    }
+  }
+
   /**
     * Transforms a stream using a single value function from original stream
     * input to new stream emission
     * @param f the function to use to transform values
     * @tparam O the type of values in the new stream
     * @return a stream with values transformed by the given function
     */
-  def map[O](f: T => O): Stream[O] = {
-    val ret = new Stream[O] {
-      private val parent = self
-      override val expectedPeriodicity = self.expectedPeriodicity
-      override val originTimeStream = self.originTimeStream
-    }
+  def map[O](f: T => O, allowRestart: Boolean = true): Stream[O] = {
+    new Stream[O] {
+      var unsubscribe: Cancel = null
 
-    val ptr = WeakReference(ret)
+      override def subscribeToParents(): Unit = {
+        unsubscribe = self.foreach(v => this.publishValue(f(v)))
+      }
 
-    var cancel: Cancel = null
-    cancel = this.foreach { v =>
-      ptr.get match {
-        case Some(s) =>
-          s.publishValue(f(v))
-        case None =>
-          cancel.apply()
+      override def unsubscribeFromParents(): Unit = {
+        unsubscribe.cancel(); unsubscribe = null
+      }
+
+      override def checkRelaunch(): Unit = {
+        if (!allowRestart) {
+          throw new IllegalStateException("This mapped stream has been marked as non-relaunchable. You may want to use .preserve to prevent stream shutdown.")
+        }
```

Comment:
It looks like this is the only place where `allowRestart` is used.
If so, what is the point of `allowRestart`?

Reply:
A couple stream methods, like relativize and scanLeft, use map but have external history and so cannot be relaunched. This makes sure that those streams let the user know that they cannot be relaunched.